network: automatically retry binding the listener socket

nagisa opened the pull request:
It is possible that binding to a socket will fail at the program startup, for example because DHCP has not fetched the IP address for the interface, or the interface is down (because WiFi).

This changes the `Network` class to implement retries every 5 seconds if the initial bind has failed. Based on my experiments once the first bind succeeds, the socket will remain properly bound and working even if the interface loses the IP address or disappears entirely, so that is nice, but unknown how reliable this behaviour is. In particular my intuition was screming to me that `accept` call would raise an error which we would use to set `self.__bind_socket = None`, so that we'd know we need to rebind later.

While here I had to make some changes to the structure, for example:

* I had to replace `self.__running: bool` with `self.__stop:
threading.Event`, as that was required to have an interruptible sleepin the `__get_bound_socket` function.
* I replaced the "self-connect" with socket shutdown in order to advance `accept` loop, as there isn't a guarantee that the network was listening on `127.0.0.1` in the first place!
* I also made the SSL context creation part of the very same retry logic – now if somebody configures invalid SSL certificates, or if those certificates are made available at a later date (e.g. fetched from letsencrypt?) the velbustcp server won’t get stuck in a similar situation as it was getting stuck with socket binds.

Fixes #13

nagisa committed:
Fix JSON syntax in the settings template

nagisa committed:
network: automatically retry binding the listener socket

It is possible that binding to a socket will fail at the program
startup, for example because DHCP has not fetched the IP address for the
interface, or the interface is down (because WiFi).

This changes the `Network` class to implement retries every 5 seconds if
the initial bind has failed. Based on my experiments once the first bind
succeeds, the socket will remain properly bound and working even if the
interface loses the IP address or disappears entirely, so that is nice,
but unknown how reliable this behaviour is. In particular my intuition
was screming to me that `accept` call would raise an error which we
would use to set `self.__bind_socket = None`, so that we'd know we need
to rebind later.

While here I had to make some changes to the structure, for example:

* I had to replace `self.__running: bool` with `self.__stop:
  threading.Event`, as that was required to have an interruptible sleep
  in the `__get_bound_socket` function.
* I replaced the "self-connect" with socket shutdown in order to advance
  `accept` loop, as there isn't a guarantee that the network was
  listening on `127.0.0.1` in the first place!
* I also made the SSL context creation part of the very same retry logic
  – now if somebody configures invalid SSL certificates, or if those
  certificates are made available at a later date (e.g. fetched from
  letsencrypt?) the velbustcp server won’t get stuck in a similar
  situation as it was getting stuck with socket binds.

nagisa committed:
network: on linux set the IP_FREEBIND socket option

This negates the need to implement a retry loop in the application logic
on that system. Other benefits include a faster reaction to changes on
the interfaces (e.g. retry loop can take up-to 5 seconds to successfully
bind after a valid address has been assigned to an interface.)

nagisa commented:
Please run the github workflows before merging – I was not able to run the tests locally due to tox getting into an infinite loop on my system for some reason.

nagisa commented:
I'm also okay with removing the last commit as it is only a minor optimization that's not really necessary to achieve the fix in this PR.

nagisa commented:
r? @pvanloo

dukeofphilberg commented:
> r? @pvanloo

Hi, we'll look to review this soon. Things have been busy here at the office 😄

dukeofphilberg commented on src/velbustcp/lib/connection/tcp/network.py:
```diff
@@ -2,6 +2,7 @@
 import ssl
 import logging
+import platform
 from typing import List
```
```suggestion
from typing import List, Optional
```

dukeofphilberg commented on src/velbustcp/lib/connection/tcp/network.py:
```diff
@@ -59,39 +65,92 @@ def send(self, data: bytearray) -> None:
+                    context.load_cert_chain(self.__options.cert, keyfile=self.__options.pk)
+                    self.__context = context
+                except Exception as e:
+                    self.__logger.error("Could not initialize SSL for %s: %s", self.__options.address, e)
```
Shorthand to automatically add stacktrace to the logging call. Check https://docs.python.org/3/library/logging.html#logging.Logger.exception

```suggestion
                    self.__logger.exception("Could not initialize SSL for %s", self.__options.address)
```

dukeofphilberg commented on src/velbustcp/lib/connection/tcp/network.py:
```diff
@@ -59,39 +65,92 @@ def send(self, data: bytearray) -> None:
-                client_socket, address = self.__bind_socket.accept()
+                sock.bind(self.__options.address)
+            except OSError as e:
+                self.__logger.error("Could not bind to %s: %s", self.__options.address, e)
```
```suggestion
                self.__logger.exception("Could not bind to %s", self.__options.address)
```

dukeofphilberg commented on src/velbustcp/lib/connection/tcp/network.py:
```diff
@@ -59,39 +65,92 @@ def send(self, data: bytearray) -> None:
+            try:
+                sock.listen(0)
+            except OSError as e:
+                self.__logger.error("Could not listen on %s: %s", self.__options.address, e)
```
```suggestion
                self.__logger.exception("Could not listen on %s", self.__options.address)
```

dukeofphilberg commented on src/velbustcp/lib/connection/tcp/network.py:
```diff
@@ -59,39 +65,92 @@ def send(self, data: bytearray) -> None:
+            try:
+                client_socket, address = self.__get_bound_socket().accept()
+            except OSError as e:
+                self.__logger.error("Couldn't accept socket: %s", e)
```
```suggestion
                self.__logger.exception("Couldn't accept socket")
```

dukeofphilberg commented on src/velbustcp/lib/connection/tcp/network.py:
```diff
@@ -59,39 +65,92 @@ def send(self, data: bytearray) -> None:
-    def __accept_sockets(self) -> None:
-        """Accepts clients from socket, if the tcp server is closed it will also close socket
-        """
+    def __get_bound_socket(self) -> socket.socket:
```
```suggestion
    def __get_bound_socket(self) -> Optional[socket.socket]:
```

If we didn't start the server yet, this call will return None

dukeofphilberg reviewed (changes requested):
Hi, sorry again that this took so long to get to!

It looks and functions well, thanks for the contribution. I've added some remarks. I can run tox again through GH actions after you make these changes, the initial workflow run request has disappeared.

nagisa committed:
Update src/velbustcp/lib/connection/tcp/network.py

Co-authored-by: Philip Vanloo <26272906+pvanloo@users.noreply.github.com>

nagisa committed:
Update src/velbustcp/lib/connection/tcp/network.py

Co-authored-by: Philip Vanloo <26272906+pvanloo@users.noreply.github.com>

nagisa committed:
Update src/velbustcp/lib/connection/tcp/network.py

Co-authored-by: Philip Vanloo <26272906+pvanloo@users.noreply.github.com>

nagisa committed:
Update src/velbustcp/lib/connection/tcp/network.py

Co-authored-by: Philip Vanloo <26272906+pvanloo@users.noreply.github.com>

nagisa committed:
Update src/velbustcp/lib/connection/tcp/network.py

Co-authored-by: Philip Vanloo <26272906+pvanloo@users.noreply.github.com>

nagisa committed:
Update src/velbustcp/lib/connection/tcp/network.py

Co-authored-by: Philip Vanloo <26272906+pvanloo@users.noreply.github.com>

nagisa commented:
Thanks for the review. I have applied all of the suggestions.

dukeofphilberg commented:
> Thanks for the review. I have applied all of the suggestions.

I've started the tox gh action check but after 30minutes they were still running. I'm assuming the same issue you had locally. I'll check this again tomorrow.

nagisa committed:
network: check for __get_bound_socket returning a None

This can happen if `self.is_active()` becomes None during a call to a
`__get_bound_socket`.

nagisa committed:
network: synchronize __bind_socket modifications

It was possible for `is_active()` to become false while the
`__bind_socket` was being set up. This could lead to the server thread
getting stuck waiting to accept connections while the `stop()` function
is already trying to `join()` the server thread.

nagisa commented:
> I've started the tox gh action check but after 30minutes they were still running. I'm assuming the same issue you had locally. I'll check this again tomorrow.

In my case it was tox itself hanging, but I was able to run pytest locally and was able to reproduce the issue that GitHub hung on as well. I have pushed two commits to fix the issues pointed out by pytest.

---

The problem was that in the network.py `stop()` function we had this;

```
try:
    self.__bind_socket.shutdown(socket.SHUT_RDWR)
    self.__bind_socket.close()
    self.__bind_socket = None
except AttributeError:
    pass
```

and then 

```
self.__server_thread.join()
```

but it was possible that the `__bind_socket` gets set to a bound socket in between these two blocks, which would lead to the server thread listening on a properly functioning socket without any intent to shut down, while the `stop()` was thinking it did everything to terminate the sockets and was waiting for the server thread to shut down.

dukeofphilberg commented:
> > I've started the tox gh action check but after 30minutes they were still running. I'm assuming the same issue you had locally. I'll check this again tomorrow.
> 
> In my case it was tox itself hanging, but I was able to run pytest locally and was able to reproduce the issue that GitHub hung on as well. I have pushed two commits to fix the issues pointed out by pytest.
> 
> The problem was that in the network.py `stop()` function we had this;
> 
> ```
> try:
>     self.__bind_socket.shutdown(socket.SHUT_RDWR)
>     self.__bind_socket.close()
>     self.__bind_socket = None
> except AttributeError:
>     pass
> ```
> 
> and then
> 
> ```
> self.__server_thread.join()
> ```
> 
> but it was possible that the `__bind_socket` gets set to a bound socket in between these two blocks, which would lead to the server thread listening on a properly functioning socket without any intent to shut down, while the `stop()` was thinking it did everything to terminate the sockets and was waiting for the server thread to shut down.


tox has run now, some errors in flake8 and mypy

nagisa committed:
chore: fix flake8

nagisa committed:
chore: fix mypy concerns

nagisa commented:
Sorry about that! I have fixed both flake8’s and mypy’s concerns in the last two commits.

nagisa commented:
Just in case this fell through the cracks, is there any chance you could give this another look?

dukeofphilberg commented:
We'll be looking to get this merged this week.

Commit:
Catch ssl client err
Handle shutdown exception on listen socket
Clean up test

dukeofphilberg commented:
Hi @nagisa, thanks for your patience. I've finalized the PR with some additional logic to catch errors. Thanks again!
